cypress-io/cypress
# fix: avoid overflow issue in Firefox

- Closes https://github.com/cypress-io/cypress/issues/21881
- Closes https://github.com/cypress-io/cypress/issues/22610

### User facing changelog

The Spec Runner will no longer overflow unexpectedly in Firefox in certain situations.

### Additional details

It appears that what's really going on here is that Firefox is basing the `scrollHeight` of the container for the panels off of the initial height of the AUT before it gets scaled. So even though everything looks correct, a hidden scrollable area is created at the bottom of the window. Other browsers do not seem to calculate the scroll height this way, and this can't be triggered in any other browser. Firefox itself returns to the correct visual state after some window resizing.

Most of the time this is fine, users can't scroll to that area accidentally, due to `overflow: hidden`. What's surprising is when you hover over certain parts of the command log, it causes FF to scroll the panels area to the bottom for some reason. This can happen without devtools open if users make their window small enough.

This PR removes the AUT container `div` from its place in the document flow using `position: absolute`, which fixes this issue in Firefox as there's never any possibility this element would take up its original space. Due to the AUT scaling logic, the AUT viewport never exceeds the size of the window and always takes up the right amount of space anyway.

It's hard to pin down the exact reason this regressed from 9.x - the DOM surrounding the AUT is very different in 10.x, with a grid and many other layers.

### Steps to test

- Open any Component or E2E spec in Firefox and reduce the height of the window so that the AUT scales down significantly. 
- Run the spec, and hover on the commands
- Observe that the UI does not shift, whereas before this change, it might have
- Use `cy.screenshot()` and confirm the screenshots look as you expect. I've tested and found no problems, all the tests and screenshot diffs themselves still pass, but every time we touch this area, we should double check that nothing has been disturbed.

### How has the user experience changed?

The state below is no longer possible to reach in FF:

<img width="1241" alt="Screen Shot 2022-05-02 at 8 51 54 AM" src="https://user-images.githubusercontent.com/1271364/171217006-32b9adb9-c1d6-40e6-baa9-ad727cd765e3.png">


### PR Tasks


- [na] Have tests been added/updated? - Cypress-in-Cypress tests cannot run in Firefox
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
